### PR TITLE
fork dl

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

File: jasmine_shim.ts [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        for (const key in { ...a, ...b }) {
+          if (!compare(a[key], b[key])) {
+            return false;
+          }
+        }
+        seen.set(a, b);
+        return true;
+      }
+      return false;
+    })(value, expected);
```
Comment (ry) [inline jasmine_shim.ts]: This looks useful. Maybe this logic should be in util.ts? Our current equality doesn't handle strings.

File: jasmine_shim.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      }
+      if (isNaN(a) && isNaN(b)) {
+        return true;
+      }
+      if (a && typeof a === "object" && b && typeof b === "object") {
+        if (seen.get(a) === b) {
+          return true;
+        }
+        for (const key in { ...a, ...b }) {
+          if (!compare(a[key], b[key])) {
```
Comment (ry) [inline jasmine_shim.ts]: impressive

File: util.ts [hunk 3]
```diff
@@ -20,6 +20,7 @@ const debug = false;
 const J = JSON.stringify;
 export const IS_WEB = typeof window !== "undefined";
 export const IS_NODE = !IS_WEB;
+export const global = (0, eval)(IS_WEB ? "window" : "global");
```
Comment (ry) [inline util.ts]: Can you avoid this comma operator? It's rather obtuse.

Comment (piscisaureus) [reply util.ts]: done

File: jasmine_shim.ts (hunk 1 above)
Comment (piscisaureus) [reply jasmine_shim.ts]: This can be done later.

File: test_browser.ts [hunk 4]
```diff
@@ -36,7 +36,7 @@ const TESTS = [
   { href: "notebook_mnist.html" }
 ];
 
-const debug = !!process.env.PP_DEBUG;
+const debug = !!process.env.PP_TEST_DEBUG;
```
Comment (ry) [inline test_browser.ts]: Can you add a comment explaining what this does

File: test_browser.ts [hunk 5]
```diff
@@ -37,6 +37,15 @@ const TESTS = [
 ];
 
 const debug = !!process.env.PP_TEST_DEBUG;
+const testdl = !!process.env.PP_TEST_DL;
```
Comment (ry) [inline test_browser.ts]: Maybe document both of these env vars at the top of the file?

File: test_browser.ts (hunk 4 above)
Comment (piscisaureus) [reply test_browser.ts]: done

**Pull request merged**